// svix/dist/openapi/models/ObjectSerializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectSerializer = void 0;
__exportStar(require("./ApplicationIn"), exports);
__exportStar(require("./ApplicationOut"), exports);
__exportStar(require("./DashboardAccessOut"), exports);
__exportStar(require("./EndpointIn"), exports);
__exportStar(require("./EndpointMessageOut"), exports);
__exportStar(require("./EndpointOut"), exports);
__exportStar(require("./EndpointSecretOut"), exports);
__exportStar(require("./EndpointStats"), exports);
__exportStar(require("./EventTypeIn"), exports);
__exportStar(require("./EventTypeOut"), exports);
__exportStar(require("./EventTypeUpdate"), exports);
__exportStar(require("./HTTPValidationError"), exports);
__exportStar(require("./HttpErrorOut"), exports);
__exportStar(require("./ListResponseApplicationOut"), exports);
__exportStar(require("./ListResponseEndpointMessageOut"), exports);
__exportStar(require("./ListResponseEndpointOut"), exports);
__exportStar(require("./ListResponseEventTypeOut"), exports);
__exportStar(require("./ListResponseMessageAttemptEndpointOut"), exports);
__exportStar(require("./ListResponseMessageAttemptOut"), exports);
__exportStar(require("./ListResponseMessageEndpointOut"), exports);
__exportStar(require("./ListResponseMessageOut"), exports);
__exportStar(require("./MessageAttemptEndpointOut"), exports);
__exportStar(require("./MessageAttemptOut"), exports);
__exportStar(require("./MessageEndpointOut"), exports);
__exportStar(require("./MessageIn"), exports);
__exportStar(require("./MessageOut"), exports);
__exportStar(require("./MessageStatus"), exports);
__exportStar(require("./ValidationError"), exports);
const ApplicationIn_1 = require("./ApplicationIn");
const ApplicationOut_1 = require("./ApplicationOut");
const DashboardAccessOut_1 = require("./DashboardAccessOut");
const EndpointIn_1 = require("./EndpointIn");
const EndpointMessageOut_1 = require("./EndpointMessageOut");
const EndpointOut_1 = require("./EndpointOut");
const EndpointSecretOut_1 = require("./EndpointSecretOut");
const EndpointStats_1 = require("./EndpointStats");
const EventTypeIn_1 = require("./EventTypeIn");
const EventTypeOut_1 = require("./EventTypeOut");
const EventTypeUpdate_1 = require("./EventTypeUpdate");
const HTTPValidationError_1 = require("./HTTPValidationError");
const HttpErrorOut_1 = require("./HttpErrorOut");
const ListResponseApplicationOut_1 = require("./ListResponseApplicationOut");
const ListResponseEndpointMessageOut_1 = require("./ListResponseEndpointMessageOut");
const ListResponseEndpointOut_1 = require("./ListResponseEndpointOut");
const ListResponseEventTypeOut_1 = require("./ListResponseEventTypeOut");
const ListResponseMessageAttemptEndpointOut_1 = require("./ListResponseMessageAttemptEndpointOut");
const ListResponseMessageAttemptOut_1 = require("./ListResponseMessageAttemptOut");
const ListResponseMessageEndpointOut_1 = require("./ListResponseMessageEndpointOut");
const ListResponseMessageOut_1 = require("./ListResponseMessageOut");
const MessageAttemptEndpointOut_1 = require("./MessageAttemptEndpointOut");
const MessageAttemptOut_1 = require("./MessageAttemptOut");
const MessageEndpointOut_1 = require("./MessageEndpointOut");
const MessageIn_1 = require("./MessageIn");
const MessageOut_1 = require("./MessageOut");
const ValidationError_1 = require("./ValidationError");
let primitives = [
    "string",
    "boolean",
    "double",
    "integer",
    "long",
    "float",
    "number",
    "any"
];
const supportedMediaTypes = {
    "application/json": Infinity,
    "application/octet-stream": 0
};
let enumsMap = new Set([
    "MessageStatus",
]);
let typeMap = {
    "ApplicationIn": ApplicationIn_1.ApplicationIn,
    "ApplicationOut": ApplicationOut_1.ApplicationOut,
    "DashboardAccessOut": DashboardAccessOut_1.DashboardAccessOut,
    "EndpointIn": EndpointIn_1.EndpointIn,
    "EndpointMessageOut": EndpointMessageOut_1.EndpointMessageOut,
    "EndpointOut": EndpointOut_1.EndpointOut,
    "EndpointSecretOut": EndpointSecretOut_1.EndpointSecretOut,
    "EndpointStats": EndpointStats_1.EndpointStats,
    "EventTypeIn": EventTypeIn_1.EventTypeIn,
    "EventTypeOut": EventTypeOut_1.EventTypeOut,
    "EventTypeUpdate": EventTypeUpdate_1.EventTypeUpdate,
    "HTTPValidationError": HTTPValidationError_1.HTTPValidationError,
    "HttpErrorOut": HttpErrorOut_1.HttpErrorOut,
    "ListResponseApplicationOut": ListResponseApplicationOut_1.ListResponseApplicationOut,
    "ListResponseEndpointMessageOut": ListResponseEndpointMessageOut_1.ListResponseEndpointMessageOut,
    "ListResponseEndpointOut": ListResponseEndpointOut_1.ListResponseEndpointOut,
    "ListResponseEventTypeOut": ListResponseEventTypeOut_1.ListResponseEventTypeOut,
    "ListResponseMessageAttemptEndpointOut": ListResponseMessageAttemptEndpointOut_1.ListResponseMessageAttemptEndpointOut,
    "ListResponseMessageAttemptOut": ListResponseMessageAttemptOut_1.ListResponseMessageAttemptOut,
    "ListResponseMessageEndpointOut": ListResponseMessageEndpointOut_1.ListResponseMessageEndpointOut,
    "ListResponseMessageOut": ListResponseMessageOut_1.ListResponseMessageOut,
    "MessageAttemptEndpointOut": MessageAttemptEndpointOut_1.MessageAttemptEndpointOut,
    "MessageAttemptOut": MessageAttemptOut_1.MessageAttemptOut,
    "MessageEndpointOut": MessageEndpointOut_1.MessageEndpointOut,
    "MessageIn": MessageIn_1.MessageIn,
    "MessageOut": MessageOut_1.MessageOut,
    "ValidationError": ValidationError_1.ValidationError,
};
class ObjectSerializer {
    static findCorrectType(data, expectedType) {
        if (data == undefined) {
            return expectedType;
        }
        else if (primitives.indexOf(expectedType.toLowerCase()) !== -1) {
            return expectedType;
        }
        else if (expectedType === "Date") {
            return expectedType;
        }
        else {
            if (enumsMap.has(expectedType)) {
                return expectedType;
            }
            if (!typeMap[expectedType]) {
                return expectedType;
            }
            let discriminatorProperty = typeMap[expectedType].discriminator;
            if (discriminatorProperty == null) {
                return expectedType;
            }
            else {
                if (data[discriminatorProperty]) {
                    var discriminatorType = data[discriminatorProperty];
                    if (typeMap[discriminatorType]) {
                        return discriminatorType;
                    }
                    else {
                        return expectedType;
                    }
                }
                else {
                    return expectedType;
                }
            }
        }
    }
    static serialize(data, type, format) {
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.serialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            if (format == "date") {
                let month = data.getMonth() + 1;
                month = month < 10 ? "0" + month.toString() : month.toString();
                let day = data.getDate();
                day = day < 10 ? "0" + day.toString() : day.toString();
                return data.getFullYear() + "-" + month + "-" + day;
            }
            else {
                return data.toISOString();
            }
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            type = this.findCorrectType(data, type);
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            let instance = {};
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.baseName] = ObjectSerializer.serialize(data[attributeType.name], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static deserialize(data, type, format) {
        type = ObjectSerializer.findCorrectType(data, type);
        if (data == undefined) {
            return data;
        }
        else if (primitives.indexOf(type.toLowerCase()) !== -1) {
            return data;
        }
        else if (type.lastIndexOf("Array<", 0) === 0) {
            let subType = type.replace("Array<", "");
            subType = subType.substring(0, subType.length - 1);
            let transformedData = [];
            for (let index in data) {
                let date = data[index];
                transformedData.push(ObjectSerializer.deserialize(date, subType, format));
            }
            return transformedData;
        }
        else if (type === "Date") {
            return new Date(data);
        }
        else {
            if (enumsMap.has(type)) {
                return data;
            }
            if (!typeMap[type]) {
                return data;
            }
            let instance = new typeMap[type]();
            let attributeTypes = typeMap[type].getAttributeTypeMap();
            for (let index in attributeTypes) {
                let attributeType = attributeTypes[index];
                instance[attributeType.name] = ObjectSerializer.deserialize(data[attributeType.baseName], attributeType.type, attributeType.format);
            }
            return instance;
        }
    }
    static normalizeMediaType(mediaType) {
        if (mediaType === undefined) {
            return undefined;
        }
        return mediaType.split(";")[0].trim().toLowerCase();
    }
    static getPreferredMediaType(mediaTypes) {
        if (!mediaTypes) {
            return "application/json";
        }
        const normalMediaTypes = mediaTypes.map(this.normalizeMediaType);
        let selectedMediaType = undefined;
        let selectedRank = -Infinity;
        for (const mediaType of normalMediaTypes) {
            if (supportedMediaTypes[mediaType] > selectedRank) {
                selectedMediaType = mediaType;
                selectedRank = supportedMediaTypes[mediaType];
            }
        }
        if (selectedMediaType === undefined) {
            throw new Error("None of the given media types are supported: " + mediaTypes.join(", "));
        }
        return selectedMediaType;
    }
    static stringify(data, mediaType) {
        if (mediaType === "application/json") {
            return JSON.stringify(data);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.stringify.");
    }
    static parse(rawData, mediaType) {
        if (mediaType === undefined) {
            throw new Error("Cannot parse content. No Content-Type defined.");
        }
        if (mediaType === "application/json") {
            return JSON.parse(rawData);
        }
        throw new Error("The mediaType " + mediaType + " is not supported by ObjectSerializer.parse.");
    }
}
exports.ObjectSerializer = ObjectSerializer;
